File: Programmers/LEVEL 2/프린터.cpp
#include <string>
#include <utility>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> priorities, int location) {
    priority_queue<int> pq;
    queue<int> q;    
    int answer = 0;
    
    for (int i = 0; i < priorities.size(); i++) {
        pq.push(priorities[i]);
        q.push(i);
    }
    
    while (!pq.empty()) {
        int index = q.front();
        int p = priorities[q.front()];
        q.pop();
        
        if (p == pq.top()) {
            pq.pop();
            answer++;
            
            if (location == index)
                break;
        }  
        else
            q.push(index);
    }
    
    return answer;
}

File: Programmers/난이도모음/LEVEL 2/프린터.java
import java.util.Queue;
import java.util.PriorityQueue;
import java.util.LinkedList;
import java.util.Collections;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        Queue<Integer> q = new LinkedList<>();
        
        for (int i = 0; i < priorities.length; ++i) {
            pq.add(priorities[i]);
            q.add(i);
        }
        
        while (true) {
            if (pq.isEmpty())
                break;
            
            int index = q.peek();
            int pick = priorities[index];
            q.poll();
            
            if (pick == pq.peek()) {
                pq.poll();
                answer++;
                
                if (location == index)
                    break;
            }
            else {
                q.add(index);
            }
            
        }
        
        return answer;
    }
}